examples/dom/WebReconciler.re

```diff
@@ -0,0 +1,229 @@
+          let node: Js.t(Dom_html.imageElement) = Obj.magic(n);
+          node##.innerHTML := Js.string(text);
+          node |> Dom_html.element;
+        },
+        children,
+      }
+    );
+  };
+
+  let createElement = (~text, ~children, ()) => {
+    JsooReact.element(make(~text, JsooReact.listToElement(children)));
+  };
+};
+
+module Button = {
+  let component = JsooReact.nativeComponent("Button");
+
+  let make = (~onPress, ~title, children) => {
+    component((_: JsooReact.Hooks.empty) =>
+      {
+        make: () => {
+          let node =
+            Dom_html.createButton(~_type=Js.string("button"), document);
+          let t = Js.string(title);
+          node##.title := t;
+          node##.innerHTML := t;
+          node##.onclick :=
+            Dom_html.handler(_e => {
+              onPress();
+              Js.bool(false);
+            });
+          node |> Dom_html.element;
+        },
+        configureInstance: (~isFirstRender as _, n) => {
+          /* TODO: Proper way to downcast? */
+          let node: Js.t(Dom_html.imageElement) = Obj.magic(n);
+          let t = Js.string(title);
+          node##.title := t;
+          node##.innerHTML := t;
+          node##.onclick :=
+            Dom_html.handler(_e => {
+              onPress();
+              Js.bool(false);
+            });
+          node |> Dom_html.element;
+        },
+        children,
+      }
+    );
+  };
+
+  let createElement = (~onPress, ~title, ~children, ()) => {
+    JsooReact.element(
+      make(~onPress, ~title, JsooReact.listToElement(children)),
+    );
+  };
+};
+
+/*
+    Step 4: Use the reconciler + native elements to create something great!
+ */
+type action =
+  | Increment
+  | Decrement;
+
+let reducer = (action, state) =>
+  switch (action) {
+  | Increment => state + 1
+  | Decrement => state - 1
+  };
+
+module CounterButtons = {
+  let component = JsooReact.component("CounterButtons");
+
+  let make = () =>
+    component(slots => {
+      let (count, dispatch, _slots: JsooReact.Hooks.empty) =
+        JsooReact.Hooks.reducer(~initialState=0, reducer, slots);
+      <View>
+        <Button title="Decrement" onPress={() => dispatch(Decrement)} />
+        <Text text={"Counter: " ++ str(count)} />
+        <Button title="Increment" onPress={() => dispatch(Increment)} />
+      </View>;
+    });
+
+  let createElement = (~children as _, ()) => JsooReact.element(make());
+};
+
+let render = () =>
+  <View> <Text text="Hello World" /> <CounterButtons /> </View>;
+
+/*
+    Step 5: Make the first render
+ */
+let rendered =
+  ref(
+    JsooReact.RenderedElement.render(
+      Dom_html.getElementById_exn("app"),
+      render(),
+    ),
+  );
+EventDom.subscribe(
+  Reconciler.onStale,
+  () => {
+    let nextElement =
+      JsooReact.RenderedElement.flushPendingUpdates(rendered^);
+    JsooReact.RenderedElement.executeHostViewUpdates(rendered^) |> ignore;
+    rendered := nextElement;
```

Comment:
clever, awesome.

Reply:
@bryphe and Revery's Codebase ™️  Sourcing Ideas Since 2018.